lib/webtask.js

```diff
@@ -665,6 +674,45 @@ exports.ensure_authenticated = function(ctx) {
                         else if (req.claims.pctx && req.claims.pctx['webtask_' + claim])
                             req.claims[claim] = req.claims.pctx['webtask_' + claim];
                     });
+
+                    if(ectx && ectx.JWT_SECRET) {
+                        if(!jwt.payload.auth)
+                            jwt.payload.auth = {};
+
+                        jwt.payload.auth.secret = ectx.JWT_SECRET;
+
+                        delete ectx.JWT_SECRET;
+                    }
+                    
+                    if(jwt.payload.auth) {
+                        if(!req.headers['authorization'] && !req.query.key) {
+                            msg = { code: 401, message: 'must provide authorization header or key query', req_id: req.req_id };
+                            ctx.logger.error({
+                                req_id: req.req_id,
+                                path: req.originalUrl,
+                                method: req.method,
+                                keyset: jwt.header.kid,
+                                payload: jwt.payload,
+                                credential: credential
+                            }, msg.message);
+                            return callback(msg);
+                        }
+
+                        try {
+                            jwt.payload.auth.user = validate_auth_token(jwt.payload.auth);
+                        } catch(msg) {
+                            ctx.logger.error({
+                                req_id: req.req_id,
+                                path: req.originalUrl,
+                                method: req.method,
+                                keyset: jwt.header.kid,
+                                payload: jwt.payload,
+                                credential: credential
+                            }, msg.message);
```

Comment:
The `must provide authorization header or key query` likely does not capture the actual issue if the blob failed validation. Either way we should probably provide more accurate error message for the different failure conditions. Lastly, the `must provide` message itself could be refactored a bit to align with the esthetics/tense etc. of other error messages we return. 

Error messages are really part of public API, and a great opportunity to educate the user (even if it is just a fellow dev on the team). Let's max out their utility.